File: jacobi-mpi.f90
program jacobi

  use mpi
  implicit none

  real(8), parameter :: top=1.0,bottom=10.0, left=1.0,right=1.0
  integer, parameter :: max_iter=100000
  integer, parameter :: nprint=100
  integer :: nx,ny
  integer :: i,j,k,iter

  real(8) :: norm,bnorm,tmpnorm,rnorm
  real(8) :: tol=1e-4
  real(8), allocatable, dimension(:,:) :: grid,grid_new 
  real(8) :: start_time
  character(len=32) :: arg

  integer :: ierr,myrank,nsize,local_nx,requests(4)

  call mpi_init(ierr)
  call mpi_comm_size(MPI_COMM_WORLD, nsize, ierr)
  call mpi_comm_rank(MPI_COMM_WORLD, myrank, ierr)

  start_time=MPI_Wtime()

  if (command_argument_count() /= 2) then
     print *, &
          "You must provide two command line arguments, the global size in X and the global size in Y"
     stop
  end if

  call get_command_argument(1, arg)
  read(arg,*) nx
  call get_command_argument(2, arg)
  read(arg,*) ny

  if (myrank==0) print *,'grid size ',nx, ' x ',ny

  local_nx=nx/nsize
  if (local_nx * nsize .lt. nx) then
      if (myrank .lt. nx - local_nx * nsize) local_nx=local_nx+1
  end if

  allocate(grid(0:ny+1,0:local_nx+1),grid_new(0:ny+1,0:local_nx+1) )


  ! boundary conditions

  grid(0,:)=left
  grid(local_nx+1,:)=right
  if (myrank ==0) then
     grid(:,0) = top
  endif
  if (myrank ==nsize-1) then
     grid(:,local_nx+1)=bottom
  endif

  ! starting conditions
  grid(1:ny,1:local_nx)=0.0
  grid_new=grid


  !! initial norm value

  tmpnorm=0.0
  do j=1,local_nx
     do i=1,ny
        tmpnorm=tmpnorm+((grid(i,j)*4-grid(i-1,j)-grid(i+1,j)-grid(i,j-1)-grid(i,j+1))**2)
     enddo
  enddo
  call mpi_allreduce(tmpnorm, bnorm, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD, ierr)


  bnorm=sqrt(bnorm)

  requests=MPI_REQUEST_NULL

  do iter=1, max_iter

     !  Insert Halo communications here
     !  HINT: You will need  two ifs, 1 for myrank>0, 1 for myrank <nsize-1
     !        Each if should have a MPI_Send/MPI_Recv pair. 
     ! 
    tmpnorm=0.0
     do j=1,local_nx
        do i=1,ny
           tmpnorm=tmpnorm+((grid(i,j)*4-grid(i-1,j)-grid(i+1,j)-grid(i,j-1)-grid(i,j+1))**2)
        enddo
     enddo

     call mpi_allreduce(tmpnorm, rnorm, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD, ierr)
     norm=sqrt(rnorm)/bnorm
 
    if (norm .lt. tol) exit

     do j=1,local_nx
        do i=1,ny
           grid_new(i,j)=0.25 * (grid(i-1,j) + grid(i+1,j) + grid(i,j-1) + grid(i,j+1))
        end do
     end do

     grid=grid_new

     if (mod(iter,nprint)==0 .and. myrank==0) then
        write(*,*) 'Iteration ',iter,' Relative norm ',norm
     endif

  enddo
  if (myrank ==0) write(*,*) 'Terminated on ',iter, ' iterations, relative norm=',norm, 'runtime=',MPI_Wtime()-start_time
  
  deallocate(grid,grid_new)

  call mpi_finalize(ierr)
end program jacobi

File: jacobi-mpi.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "mpi.h"
#include <string.h>


// Grid boundary conditions
#define RIGHT 1.0
#define LEFT 1.0
#define TOP 1.0
#define BOTTOM 10.0

// Algorithm settings
#define TOLERANCE 0.0001
#define NPRINT 1000
#define MAX_ITER 10000

int main(int argc, char*argv[]) {

  int k;
  double tmpnorm,bnorm,norm;

  int size,myrank; 
  MPI_Request requests[]={MPI_REQUEST_NULL, MPI_REQUEST_NULL, MPI_REQUEST_NULL, MPI_REQUEST_NULL};

  MPI_Init(&argc,&argv);
  MPI_Comm_size(MPI_COMM_WORLD,&size);
  MPI_Comm_rank(MPI_COMM_WORLD,&myrank);

  if (argc !=3) {
    if (myrank==0) printf("usage: %s GRIDX GRIDY\n",argv[0]);
    return(-1);
  }


  int nx=atoi(argv[1]);
  int ny=atoi(argv[2]);
  int ny2=ny+2;

  if (myrank==0) printf("grid size %d X %d \n",ny,ny);

  // local grid size
  int local_nx=nx/size;
  if (local_nx * size < nx) {
    if (myrank < nx - local_nx * size) local_nx++;
  }

  double *grid= (double*)malloc(sizeof(double)*(local_nx+2)*(ny+2));
  double *grid_new= (double*)malloc(sizeof(double)*(local_nx+2)*(ny+2));
  double start_time;
  
 
  // Initialise Grid boundaries
  int i,j;
  for (i=0;i<ny+2;i++) {
    grid_new[i]=grid[i]=  (myrank==0 ?TOP:0.0);
    j=(ny+2)*(local_nx+1)+i;
    grid_new[j]=grid[j]=  (myrank==size-1?BOTTOM:0.0);
  }
  for (i=1;i<local_nx+1;i++) {
    j=(ny+2)*i;
    grid_new[j]=grid[j]=LEFT;
    grid_new[j+ny+1]=grid[j+ny+1]=RIGHT;
  }
   
  // Initialise rest of grid
  for (i=1;i<=local_nx;i++) 
    for (j=1;j<=ny;j++)
      k=(ny+2)*i+j;
  grid_new[k]=grid[k]=0.0;
   
  // Calculate initial norm
  tmpnorm=0.0;
  for (i=1;i<=local_nx;i++) {
    for (j=1;j<=ny;j++) {
      k=(ny+2)*i+j;            
      tmpnorm=tmpnorm+pow(grid[k]*4-grid[k-1]-grid[k+1] - grid[k-(ny+2)] - grid[k+(ny+2)], 2); 

    }
  }
  MPI_Allreduce(&tmpnorm, &bnorm, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
  bnorm=sqrt(bnorm);

  start_time=MPI_Wtime();
  int iter;
  for (iter=0; iter<MAX_ITER; iter++) {

    // Halo communications
    // Insert halo communications here:	
     //  HINT: You will need  two ifs, 1 for myrank>0, 1 for myrank <nsize-1
     //        Each if should have a MPI_Send/MPI_Recv pair. 
	
    // Recalculate norm
    //
    double rnorm;
    tmpnorm=0.0;
    for (i=1;i<=local_nx;i++) {
      for (j=1;j<=ny;j++) {
	k=(ny+2)*i+j;
	tmpnorm=tmpnorm+pow(grid[k]*4-grid[k-1]-grid[k+1] - grid[k-(ny+2)] - grid[k+(ny+2)], 2); 
      }
    }

    // Sum up all the tmpnorms and copy to all nodes
    // mpi command here
    norm=sqrt(rnorm)/bnorm;

    // Convergence test
    //
    if (norm < TOLERANCE) break;

    // Calculate new grid
    for (i=1;i<=local_nx;i++) {
      for (j=1;j<=ny;j++) {
	k=(ny+2)*i+j;    
	grid_new[k]=0.25 * (grid[k-1]+grid[k+1] + grid[k-(ny+2)] + grid[k+(ny+2)]);
      }
    }

    // Copy newgrid into old one for next cycle
    memcpy(grid, grid_new, sizeof(double) * (local_nx + 2) * (ny+2));

    if (iter % NPRINT ==0 && myrank==0 ) printf("Iteration =%d ,Relative norm=%e\n",iter,norm);
  }
  // End of iterative cycle

  if (myrank==0) printf("Terminated on %d iterations, Relative Norm=%e, Total time=%e seconds \n", iter,norm, MPI_Wtime() - start_time);
  


  free(grid);
  //free(temp);
  free(grid_new);

  MPI_Finalize();
  return 0;
    

}
